File: 290. Word Pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if (pattern.length() != words.length) return false;
        Map<Character, String> mapping = new HashMap<>();
        Map<String, Character> reverseMapping = new HashMap<>();
        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            String w = words[i];
            if (mapping.containsKey(c) && !mapping.get(c).equals(w)) return false;
            if (reverseMapping.containsKey(w) && !reverseMapping.get(w).equals(c)) return false;
            mapping.put(c, w);
            reverseMapping.put(w, c);
        }
        return true;
    }
}

File: 290. Word Pattern.go
func wordPattern(pattern string, s string) bool {
    forwardMap := make(map[byte]string)
    reverseMap := make(map[string]byte)
    words := strings.Split(s, " ")
    if len(words) != len(pattern) {
        return false
    }
    for i := 0; i < len(pattern); i++ {
        char := pattern[i]
        word := words[i]

        mapWord, wordExists := forwardMap[char]
        if wordExists && mapWord != word {
            return false
        }

        mapChar, charExists := reverseMap[word]
        if charExists && mapChar != char {
            return false
        }

        forwardMap[char] = word
        reverseMap[word] = char
    }
    return true
}
